### PR TITLE
Fix: module declaration with external imported module behavior

### LINKED ISSUE #7669: [@parcel/transformer-typescript-types]: Meaningless "Got unexpected undefined" error message
I encountered "Got unexpected undefined" error. I don't know what exactly causes it, but I created repo with minimal reproduction.

## 🎛 Configuration

`package.json`
```JSON
{
  "name": "parcel-bug",
  "main": "dist/index.js",
  "source": "src/index.ts",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "parcel build"
  },
  "dependencies": {
    "next": "^12.0.10"
  },
  "devDependencies": {
    "@parcel/packager-ts": "^2.2.1",
    "@parcel/transformer-typescript-types": "^2.2.1",
    "parcel": "^2.2.1",
    "typescript": "^4.5.5"
  }
}
```

`tsconfig.json`
```JSON
{
  "extends": "./tsconfig.options.json",
  "compilerOptions": {
    "composite": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "jsx": "react-jsx",
    "lib": ["DOM", "DOM.Iterable", "ES2021"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmit": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUncheckedIndexedAccess": true,
    "rootDir": "./src",
    "strict": true,
    "target": "ES2021",
  },
  "include": ["src/**/*"],
  "references": []
}

```

Source code:
```TS
import type { NextWebVitalsMetric } from 'next/app';

declare module 'next/app' {
  export interface AppInitialProps {
    pageProps: unknown;
  }
}

export async function reportWebVitals({ value, name }: NextWebVitalsMetric) {}
```

## 🤔 Expected Behavior

Parcel builds with no error? I think this minimal reproduction shouldn't actually build (because of e.g. missing React), but the errors should be more descriptive and useful.

## 😯 Current Behavior

Parcel errors:

```
× Build failed.

@parcel/transformer-typescript-types: Got unexpected undefined

  Error: Got unexpected undefined
  at nullthrows (D:\projects\parcel-test\node_modules\nullthrows\nullthrows.js:7:15)
  at TSModuleGraph.propagate
  (D:\projects\parcel-test\node_modules\@parcel\transformer-typescript-types\lib\TSModuleGraph.js:271:48)
  at shake (D:\projects\parcel-test\node_modules\@parcel\transformer-typescript-types\lib\shake.js:42:35)
  at D:\projects\parcel-test\node_modules\@parcel\transformer-typescript-types\lib\TSTypesTransformer.js:139:33
  at transformation (D:\projects\parcel-test\node_modules\typescript\lib\typescript.js:107043:24)
  at transformRoot (D:\projects\parcel-test\node_modules\typescript\lib\typescript.js:107070:82)
  at Object.transformNodes (D:\projects\parcel-test\node_modules\typescript\lib\typescript.js:107054:78)
  at emitDeclarationFileOrBundle (D:\projects\parcel-test\node_modules\typescript\lib\typescript.js:107768:43)
  at emitSourceFileOrBundle (D:\projects\parcel-test\node_modules\typescript\lib\typescript.js:107671:13)
  at forEachEmittedFile (D:\projects\parcel-test\node_modules\typescript\lib\typescript.js:107397:30)
```

## 💻 Code Sample

[Repo with reproduction](https://github.com/jakubmazanec/parcel-bug). Clone and run `npm install` and `npm run build`.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           |  2.2.1
| Node             | 16.13.2
| npm | 8.3.1
| Operating System | Windows 10 Enterprise 20H2

### PR DESCRIPTION
**Pull request by Shinyaigeek**
Fixes: #7669 

# ↪️ Pull Request

Hi team 👋  I investigated #7669, and I noticed the cause of this issue.

https://github.com/parcel-bundler/parcel/blob/v2/packages/transformers/typescript-types/src/TSModuleGraph.js#L129-L135

In these line, if `this.getModule` returns undefined, this `resolveImport` returns `{module, name: local, imported: imported || i.imported};` because this is external module. However, if a user declares a module with the imported module specifier, like `declare module “external module”`, `this.getModule` return declared `TSModule`(not imported) and `resolveImport` try to resolve type with the declared module, not with the imported module. This causes modules declaration **fully** to overwrite imported modules.

I fixed this by passing through the import if `this.resolveExport` (with declared module) returns undefined as a fallback. I think this is ideal behavior because module declaration overrides **only** what is declared.
## 💻 Examples

```typescript

import type { hoge } from 'hoge';

declare module 'hoge' {
  export interface Bar {}
}

const h: hoge = 'hoge';

```

## 🚨 Test instructions

I wrote the test to cover this case

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by jakubmazanec** [comment]
Any idea when this will be merged and #7669 fixed?

**Comment by Shinyaigeek** [comment]
I don’t know when this will be merged. it is needed to get a review from maintainer  in order to merge this PR.

I think it can be a workaround to declare also types which is imported and used from “mod" in `declare module “mod”`.

**Comment by wbinnssmith** [comment]
Unfortunately I don't know much about our TypeScript processing — @devongovett or @DeMoorJasper would be best suited to review this.

**Comment by rathpc** [comment]
Wondering if @devongovett or @DeMoorJasper will be able to give this a look soon?

**Comment by jakubmazanec** [comment]
Almost three months and still no review. Can I do anything to help? 🙏 We really need #7669 fixed. At our company, we could benefit from Parcel's speed, but unresolved bugs like this means Parcel is unusable.

File: packages/transformers/typescript-types/src/TSModuleGraph.js [hunk 1]
```diff
@@ -127,12 +127,19 @@ export class TSModuleGraph {
...
+    };
     if (!m) {
       // External module. pass through the import.
-      return {module, name: local, imported: imported || i.imported};
+      return importedModule;
     }
 
-    return this.resolveExport(m, imported || i.imported);
+    // passing through the import as a fallback for declare module with external module specifier
+    return this.resolveExport(m, imported || i.imported) ?? importedModule;
```
Comment (devongovett) [inline packages/transformers/typescript-types/src/TSModuleGraph.js:142]: As far as I can tell when running the test this line isn't called?

**Review by devongovett — COMMENTED** [review]
Apologies for the slow review. I needed to find time to actually sit down and try this because the TS types transformer is kinda complicated.

Is it expected that this is the output?

```ts
declare module "node_modules/hoge/index" {
    interface Bar {
    }
}
```

The `node_modules/hoge/index` specifier seems kinda strange. I also get the exact same output on `v2` as well so I'm not sure if the change here is having an effect? Or maybe the test just isn't hitting it? Not sure.

**Comment by Shinyaigeek** [comment]
@devongovett sorry for the really late response… 🙇I did not have time for OSS due to my company work.

According to [this comment](https://github.com/parcel-bundler/parcel/issues/7668#issuecomment-1035260880), the base issue seems to be resolved, but an unexpected behavior still be reported (FYI: https://github.com/parcel-bundler/parcel/issues/7669#issuecomment-1106567699).

Anyway, as you mentioned, my change seems not to work in the current version of parcel,  so I will close this PR and investigate the bug again

**Pull request closed without merging**